HIP/hip-1000.md

```diff
@@ -0,0 +1,111 @@
+## User stories
+- As a primary creator of an NFT collection, I should be able to mint a large number of NFT serials in a single collection
+- As a secondary NFT marketplace dApp, I should be able to create many thousands of collections (created by multiple primary creators of the collections), such that each collection can hold a large number of NFTs
+- As a NFT creator, I should be able to set custom fees for my NFTs
+    - Is it a must-have requirement to be able to do this on per-serial number? or is it sufficient to specify this fee at the token-type level?
+- As an avid collector of NFT, I want to manage my entire collection within the same account.
+
+## Specification
+ Currently, we have a limit of 1000 token associations per Account put in place which is dictated by the bootstrap property `tokens.maxPerAccount`.
+This is enforced everytime we associate token/tokens to an account. 
+
+When using the HederaTokenStore :
+```java
+     public ResponseCodeEnum associate(AccountID aId, List<TokenID> tokens, boolean automaticAssociation) {
+      ...
+          if ((accountTokens.numAssociations() + tokenIds.size()) > properties.maxTokensPerAccount()) {
+              validity = TOKENS_PER_ACCOUNT_LIMIT_EXCEEDED;
+          }
+      ...
+    }
+```
+
+When using the Account model:
+```java
+    public void associateWith(List<Token> tokens, int maxAllowed, boolean automaticAssociation){
+        final var alreadyAssociated=associatedTokens.size();
+        final var proposedNewAssociations=tokens.size()+alreadyAssociated;
+        validateTrue(proposedNewAssociations<=maxAllowed,TOKENS_PER_ACCOUNT_LIMIT_EXCEEDED);
+        ...
+    }
+```
+
+This limit enforcement needs to be removed. This limit is also enforced when creating and updating an account with maxAutomaticAssociations described by this [HIP-23](https://github.com/anighanta/hedera-improvement-proposal/blob/master/HIP/hip-23.md)
+That enforcement also has to be removed.
+
+We currently store the list of tokens associated to an account on the state using a MerkleLeaf node which hold the array of Ids of the token and each MerkleAccount will have this as a child.
+So as the array grows it will be very costly to operate on this array. 
+
+![current.png](../assets/hip-1000/current.png)
+
+We propose a new way of efficiently doing this using a `MapValueLinkedList` which will store the Account-Token relationship as a Map of <EntityNum pair of accountID and tokenID> to <MerkleTokenRelStatus>
+where the keys will be linked by a linkedList. And only the latest account-token association key is saved on the MerkleAccount state.
+
+![proposed.png](../assets/hip-1000/proposed.png)
+
+## Backwards Compatibility
+
+There are no known backwards compatibility issues.
+
+## Security Implications
+
+This shouldn’t necessarily change security implications not already known by Hedera,
+but could introduce denial of service ? Which can be controlled by throttling and increasing fee as the associations grow. 
```

Comment:
This needs to be clarified. There is no new denial of service because an attacker today can still flood the system with accounts and associations. Creating a new account every 1,000 associations is not hard. So there is no new security implications here.